packages/dds/map/src/interfaces.ts

```diff
@@ -191,18 +190,36 @@ export interface IDirectory extends Map<string, any> {
     getWorkingDirectory(relativePath: string): IDirectory;
 }
 
+/**
+ * Interface describing actions on a directory.
+ *
+ * @remarks
+ * When used as a Map, operates on its keys.
+ */
+export interface IDirectory extends IDirectoryNoEvents, IEventProvider<IDirectoryEvents> { }
+
 export interface ISharedDirectoryEvents extends ISharedObjectEvents {
     (event: "valueChanged", listener: (
         changed: IDirectoryValueChanged,
         local: boolean,
         op: ISequencedDocumentMessage,
-        target: IEventThisPlaceHolder) => void);
+        target: IEventThisPlaceHolder,
+    ) => void);
+}
+
+export interface IDirectoryEvents extends IEvent {
+    (event: "containedValueChanged", listener: (
+        changed: IValueChanged,
+        local: boolean,
+        op: ISequencedDocumentMessage,
+        target: IEventThisPlaceHolder,
+    ) => void);
 }
 
 /**
  * Interface describing a shared directory.
  */
-export interface ISharedDirectory extends ISharedObject<ISharedDirectoryEvents>, IDirectory {
+export interface ISharedDirectory extends ISharedObject<ISharedDirectoryEvents>, IDirectoryNoEvents {
```

Comment:
this is all a bit wonky. I think we should get ride of IDirectoryNotEvents, and here you can just remove the events from IDirectory,  Omit<Omit<Omit<IDirectory,"on">,"once">,"off">. I'm also unlcear why shared directory does have containedValueChanged, shouldn't we get that on the root as well?

Reply:
Yeah I'm fighting with the typing here -- I'll try the Omit approach and see how that does.

Shared directory should have containedValueChanged -- I had ISharedDirectoryEvents extending IDirectoryEvents at one point, think I just accidentally removed that.

Reply:
yeah. it's tricky. you can do something like SharedObject<ISharedDirectoryEvents & IDirectoryEvents>, there is also the ExtendEventProvider type in common utils, but i'm not sure what fits with the composition you are doing. happy to help,

---
In reply to: [540356177](https://github.com/microsoft/FluidFramework/pull/4556#discussion_r540356177) [](ancestors = 540356177)

Reply:
if i understand correctly. you don't want shared object events on IDirectory, and you want all events on ISharedDirectory. For that i think omiting is the way. you should omit from IDirectory on ISharedDirectory, and intersect(&) the events types in the ISharedObject generic. I think that will work, and doesn't require extra types

---
In reply to: [540358116](https://github.com/microsoft/FluidFramework/pull/4556#discussion_r540358116) [](ancestors = 540358116,540356177)

Reply:
Latest iteration uses Omit but then adds back the Symbols that get stripped out, which allows us to stop exporting an intermediate no-event interface.  LMK if there are other things you'd like me to try.